### PR TITLE
Remove re-linking programs approach

### PR DESCRIPTION
**Pull request by astojilj**
Remove re-linking programs as redundant.
It costs (cheaper to link once than twice) and (subjective) is not that common GL API
usage pattern, although perfectly legal and permitted.
Initial idea, behind the removed code, was to enable work on optimization
that would reduce number of attrib setup calls in case when VAO is not
available (as described in #9433).

As such optimization is not implemented, and it is arguable if it makes sense
to do it now, we can remove re-linking.

Related to closed PRs #9433 and PR #11583.

I have [measured the time spent just on relinking](https://gist.github.com/astojilj/29bd5a5c5dc0b2d9f29ecb660da07fbf) using release build on iPhone SE (A9, same as iPhone 6S):

- 1st run after reboot or installation
Total 37.14ms, average per program:1.86ms

- reopening
Total: 2.47ms, average per program: 0.12ms

This time we save using the patch here.

### CODE REVIEW

**Comment by astojilj** [comment]
@ivovandongen @kkaefer @alexshalamov PTAL

File: src/mbgl/gl/attribute.hpp [hunk 1]
```diff
@@ -24,6 +24,7 @@ using NamedAttributeLocations = std::vector<std::pair<const std::string, Attribu
 class Context;
 void bindAttributeLocation(Context&, ProgramID, AttributeLocation, const char * name);
 std::set<std::string> getActiveAttributes(ProgramID);
```
Comment (alexshalamov) [inline src/mbgl/gl/attribute.hpp]: Is it used anywhere else?

File: src/mbgl/gl/attribute.hpp [hunk 2]
```diff
@@ -24,6 +24,7 @@ using NamedAttributeLocations = std::vector<std::pair<const std::string, Attribu
 class Context;
 void bindAttributeLocation(Context&, ProgramID, AttributeLocation, const char * name);
```
Comment (alexshalamov) [inline src/mbgl/gl/attribute.hpp]: can be removed?

File: src/mbgl/gl/attribute.cpp [hunk 3]
```diff
@@ -7,41 +7,14 @@ namespace gl {
 
 using namespace platform;
 
-void bindAttributeLocation(Context& context, ProgramID id, AttributeLocation location, const char* name) {
-    // We're using sequentially numberered attribute locations starting with 0. Therefore we can use
-    // the location as a proxy for the number of attributes.
-    if (location >= context.maximumVertexBindingCount) {
```
Comment (ivovandongen) [inline src/mbgl/gl/attribute.cpp]: Do we still have an equivalent check in place?

File: src/mbgl/gl/attribute.hpp (hunk 1 above)
Comment (ivovandongen) [reply src/mbgl/gl/attribute.hpp]: Implementation is removed, should be removed as well?

File: src/mbgl/gl/attribute.cpp (hunk 3 above)
Comment (astojilj) [reply src/mbgl/gl/attribute.cpp]: The check is done [before render](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/renderer/render_layer.cpp#L76).

File: src/mbgl/gl/attribute.hpp (hunk 2 above)
Comment (astojilj) [reply src/mbgl/gl/attribute.hpp]: Done. Thanks.

File: src/mbgl/gl/attribute.hpp (hunk 1 above)
Comment (astojilj) [reply src/mbgl/gl/attribute.hpp]: Done. Thanks. `class Context` and unnecessary includes removed, too with [patch #2](https://github.com/mapbox/mapbox-gl-native/pull/14482/commits/c1e12c80d57c6e56438a23df533db969c893df77).

File: src/mbgl/gl/attribute.cpp (hunk 3 above)
Comment (ivovandongen) [reply src/mbgl/gl/attribute.cpp]: @astojilj Correct me if I'm wrong, but the [check](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/renderer/render_layer.cpp#L76) you mention only takes care of logging, it doesn't prevent rendering. So the queryLocations is still reached. What happens when an attribute can't be found or we cross the `MAX_VERTEX_ATTRIBUTES` limit? If `glGetAttribLocation` returns -1 in that situation, things will go wrong when using the returned location as an index [here](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/gl/attribute.hpp#L88) no?

Comment (astojilj) [reply src/mbgl/gl/attribute.cpp]: No, if glGetAttribLocation returns -1, then `if (location) {` [here](https://github.com/mapbox/mapbox-gl-native/blob/858097102d9f586e17b118532213e38262e801db/src/mbgl/gl/attribute.hpp#L88), as false, guards the `result.at(*location) = binding;`.

Binding, or not binding, vertex attributes before linking, based on GL_MAX_VERTEX_ATTRIBS count, doesn't change the behavior of following glLinkProgram and [error](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glLinkProgram.xhtml).
As I saw, this was resolved by [packing attributes PR](https://github.com/mapbox/mapbox-gl-native/issues/8183) and now we use the log to inform about error.

File: src/mbgl/gl/attribute.cpp [hunk 4]
```diff
@@ -1,47 +1,19 @@
 #include <mbgl/gl/attribute.hpp>
 #include <mbgl/gl/context.hpp>
-#include <mbgl/gl/defines.hpp>
 
 namespace mbgl {
 namespace gl {
 
 using namespace platform;
 
-void bindAttributeLocation(Context& context, ProgramID id, AttributeLocation location, const char* name) {
```
Comment (kkaefer) [inline src/mbgl/gl/attribute.cpp]: This function's task is to set an explicit binding for an attribute location before (re-)linking the program. The code flow is a bit hard to follow due to it's heavily templated nature, but it basically worked like this:

- Link program
- Query program to find out what attributes are actually present. This is necessary because we disable/enable attributes for data-driven styling (DDS)
- The `AttributeLocations` "iterates" (using template expansions...) over an explicitly ordered list of attributes (they're stored in the `*_program.hpp` in a `TypeList`: https://github.com/mapbox/mapbox-gl-native/blob/8a3d07fcd83d01415b672dd5f45a5e08af7d5594/src/mbgl/programs/circle_program.hpp#L18-L19
    and augmented in the `gfx::Program` class with the attributes for data-driven styling, which is a subset of this list: https://github.com/mapbox/mapbox-gl-native/blob/8a3d07fcd83d01415b672dd5f45a5e08af7d5594/src/mbgl/style/layers/circle_layer_properties.hpp#L61-L73
  That means that attributes get sequentially numbered, starting with `location = 0` for the second linking pass. Some (many?) OpenGL implementations have massive performance issues if the attribute 0 is disabled (e.g. because it is not used). With this weird double-linking approach, we're ensuring that location 0 is always enabled because it is the first attribute of the program's non-DDS attribute list. Removing relinking is definitely possible, but we should preserve the current behavior of explicitly assigning attribute bindings with `glBindAttribLocation`. We could probably get away with only running it for the first attribute in the list, since we don't really care whether the rest are numbered sequentially. What doesn't work is to run `glBindAttribLocation` blindly for all attributes because we sometimes have more than 8 of those, and many implementations don't support attribute locations beyond 7, so we have to have some awareness of which attributes are in use or not. We could query this information from the `gfx::AttributeBindings<AttributeList>` object passed to the `draw` function (that we also use to compute the `gl::AttributeKey<AttributeList>` for finding the correct program instance in the first place). Explicitly binding all attributes before the first link also has the advantage that we don't have to query the attribute locations after linking because they're guaranteed to be the ones we set them to.

Comment (astojilj) [reply src/mbgl/gl/attribute.cpp]: > Query program to find out what attributes are actually present. This is necessary because we disable/enable attributes for data-driven styling (DDS)

> we don't have to query the attribute locations after linking because they're guaranteed to be the ones we set them to.

The patch is replacing usage of glGetProgramiv and per-attrib glGetActiveAttrib by glGetAttribLocation. I don't think it is introducing redundancy.


> With this weird double-linking approach, we're ensuring that location 0 is always enabled because it is the first attribute of the program's non-DDS attribute list. Removing relinking is definitely possible, but we should preserve the current behavior of explicitly assigning attribute bindings with glBindAttribLocation. We could probably get away with only running it for the first attribute in the list, since we don't really care whether the rest are numbered sequentially. 

This proposal sounds good and looks safe to assert (I checked the shaders code before working on this) position attribute being always enabled, for all the shaders on index 0. I'll update patch as that would avoid double linking and resolve other concerns.

Comment (kkaefer) [reply src/mbgl/gl/attribute.cpp]: > I don't think it is introducing redundancy.
I didn't mean to suggest that your patch does; I was describing the old approach.

Comment (astojilj) [reply src/mbgl/gl/attribute.cpp]: The patch [Remove re-linking: bind position attribute at location 0.](https://github.com/mapbox/mapbox-gl-native/pull/14482/commits/99e5f262645aeff19797cc1ed87b023add895e59) binds the position  related attribute at location 0. 

To expand on the comment:
    // It is important to have attribute at position 0 enabled: conveniently,
    // position attribute is always first and always enabled. The integrity of
    // this assumption is verified in AttributeLocations::queryLocations and
    // AttributeLocations::getFirstAttribName.

there is compile time check that the first element is position and assert verifying the bound location 0 enabled. 

I have [measured the time spent just on relinking](https://gist.github.com/astojilj/29bd5a5c5dc0b2d9f29ecb660da07fbf) using release build on iPhone SE (A9, same as iPhone 6S):

- 1st run after reboot or installation
Total 37.14ms, average per program:1.86ms

- reopening
Total: 2.47ms, average per program: 0.12ms

This time we save using the patch here.
